### DISCUSSION TITLE
This code results in an infinite request loop--Why?

### DISCUSSION THREAD

**Discussion by AndrewLester** [vercel/next.js#47329] [Help]
### Summary

This code results in an infinite request loop:

<img width="864" alt="a7a9dabe-121f-42e5-844e-95321849af11" src="https://user-images.githubusercontent.com/23221268/226487160-66d8401c-efb9-41f8-89e7-6155a7ee147c.png">

EDIT: Typo above, doesn't affect behavior. Should be `export default (_, res) => {`

It racked up $1600 worth of overage on Vercel before catching it. Does anyone have an idea why?

Full repro: https://github.com/AndrewLester/nextjs-repro (change `.env.production` to `.env` and change `BASE_URL` to `http://localhost:3000` if you want to test locally in dev). Deploy to Vercel to see the full effect (MAKE SURE YOU REMOVE DEPLOYMENT)

Navigate to `/` or `/api/hello`.

Preemptive docs pages:
- [Custom app `getInitialProps`](https://nextjs.org/docs/advanced-features/custom-app#:~:text=When%20you%20add%20getInitialProps%20in%20your%20custom%20app%2C%20you%20must%20import%20App%20from%20%22next/app%22%2C%20call%20App.getInitialProps(appContext)%20inside%20getInitialProps%20and%20merge%20the%20returned%20object%20into%20the%20return%20value.)
- [Server errors _should_ return a **static** page](https://nextjs.org/docs/advanced-features/error-handling#:~:text=Next.js%20provides%20a%20static%20500%20page%20by%20default%20to%20handle%20server%2Dside%20errors%20that%20occur%20in%20your%20application.)

### Additional information

```js
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.2.0: Fri Nov 11 02:04:44 PST 2022; root:xnu-8792.61.2~4/RELEASE_ARM64_T8103
    Binaries:
      Node: 18.12.1
      npm: 8.19.2
      Yarn: 1.22.19
      pnpm: 7.17.0
    Relevant packages:
      next: 13.2.4
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0

Also occurs on next 12.
```

### Example

https://github.com/AndrewLester/nextjs-repro

**Comment by icyJoseph** [comment]
Asynchronous work, ought to be handled, for both `fulfilled` and `rejected` scenarios, always.

> Previous to Node 15, unhandled rejections were just a warning, after Node 15, an unhandled rejection crashes the process. This is a measure to force developers to always handle rejection scenarios. 

```js
import '../styles/globals.css';
import App from 'next/app';

function MyApp({ Component, pageProps }) {
    return <Component {...pageProps} />;
}

MyApp.getInitialProps = async (ctx) => {
  try {
    // ⚠️ let's not forget that fetch doesn't throw if the response is not 200
   // we ought to check `response.ok` - I made an omission here earlier
   // however it'll throw if, in the server, the url is not absolute
    await fetch(`${process.env.BASE_URL}/api/hello`);

    return App.getInitialProps(ctx);
  } catch (reason) {
    if (reason instanceof Error) {
      // handle the error
      return {} //?? maybe?
    }
    // handle the unexpected exception
    return {} // ?? maybe?
  }
};

export default MyApp;
```

As, your `getInitialProps` function throws an exception, Next.js tries to generate an error page, but that's also running into the same issue, so `getInitialProps` throws yet another exception, and so on...

It also follows that you should try/catch your handler body as well, at least for sake of traceability. Though in this case, you'd still wound up in a loop I think...

```ts
export default (res) => {
  try {
    "Error".error(); // This is an accident, what can you do
    return res.status(200).json({ test: "hi" });
  } catch (reason) {
    // handle reason
    return res.status(500).json({ message: "Internal Server Error" }); // Though this will still trigger
  }
};

```

**Reply by AndrewLester** [reply]
I appreciate your help! Sorry if I seem a bit argumentative, but it's just the $1600 Vercel bill speaking. I've addressed your points below:

> As, your `getInitialProps` function throws an exception, Next.js tries to generate an error page, but that's also running into the same issue, so `getInitialProps` throws yet another exception, and so on...

RE your first point, this is not the main cause of the issue. The infinite loop starts _before_ the `getInitialProps` function's fetch exits/errors. You'll note this if you try the repro. The fetch takes a **long** time.

Secondly, it also starts erroring if I navigate directly to `/api/hello`. Why would the `getInitialProps` function run for an API handler error? The docs specifically say a ["static error"](https://nextjs.org/docs/advanced-features/error-handling#:~:text=Next.js%20provides%20a%20static%20500%20page%20by%20default%20to%20handle%20server%2Dside%20errors%20that%20occur%20in%20your%20application.) page will show for server-side errors.

Thirdly, you're insinuating that an error in `getInitialProps` causes an infinite loop on its own... that doesn't sound like good design to me! It's okay though because that's not what's happening. If you modify the repro to throw an error inside `getInitialProps`, Next.js manages to stop looping after 3 attempts of failing to render `MyApp` (at least locally).

> It also follows that you should try/catch your handler body as well, at least for sake of traceability. Though in this case, you'd still wound up in a loop I think...

Sure, I _could_ do this. The point is that the error is _unexpected_ (i.e. I can't just put a try catch around the whole handler. More likely, I _shouldn't_.).

And no, this solution _does_ actually fix the issue. Like I said before though, not feasible. I will note that I do have a typo in the repro (doesn't affect behavior), it should be:
```patch
+ export default (_, res) => { 
- export default (res) => { 
```

**Reply by icyJoseph** [reply]
That's alright. I just don't have the time to clone all repos and stuff. I have also experienced something similar with errors and App.getInitialprops, previously, which is why my first instinct was to tackle that bit.

I cloned the repo and out of the box I get an error that's true, but that's it, no infinite loop. Only in dev mode, I get the infinite loop, after creating `.env.development`.

The issue might be with the API response timing out then, because it never resolves? Still not sure what's up there. I can't really tell from building it locally. I placed a good ol' console log inside GIP and it prints only once, no issues follow.

- Are you sure the VERCEL_URL is enabled on your project?
- Is the reproduction demo complete? 

> Thirdly, you're insinuating that an error in getInitialProps causes an infinite loop on its own... 

Well, one can bend the framework till its out of its design form, and these things happen.

> And no, this solution does actually fix the issue ([500 response codes do not cause fetch to error](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#:~:text=The%20Promise%20returned%20from%20fetch()%20won%27t%20reject%20on%20HTTP%20error%20status%20even%20if%20the%20response%20is%20an%20HTTP%20404%20or%20500.)).

Right,  I rushed a bit there, thinking the problem was fetch not even flying the request for absolute URLs. You should still try/catch around that though. I imagine you are just getting a cookie on the client from that call or something like that.

**Reply by AndrewLester** [reply]
No worries, again, I appreciate your help! I suppose I should be more clear with the repro, it's meant to deploy to Vercel (hence why I have a `.env.production`), since that's originally how I came across the error. To observe locally, you must rename the `.env.production` to `.env` or `.env.development` and change `BASE_URL` to `http://localhost:3000`.

- Are you sure the VERCEL_URL is enabled on your project?
Yes (if `VERCEL_URL`) is disabled the `fetch` will fail and no loop happens.
- Is the reproduction demo complete?
Yes (it shows same behavior)

> You should still try/catch around that though.

Yeah I should, the reason I left it out for this example is because I just wanted to keep it short.

> The issue might be with the API response timing out then, because it never resolves? 

The console becomes spammed with `/api/hello` requests in an infinite loop though... after only one request to `/api/hello`. Further, if you add a `console.log(...)` to `_app.js` (ABOVE THE FETCH), that logs infinitely very quickly too! My idea:
Since `/api/hello` errors out, a `/_error` page is rendered which then runs the `_app.js` `getInitialProps`. This is **undocumented** (API handler errors shouldn't run `_app.js`, right?) behavior, yet it appears to be happening nonetheless!

**Reply by icyJoseph** [reply]
Strange, I did change the Vercel Url to localhost:3000 and run it on build mode, it went fine.

Running it in dev mode does loop though. I'll look into more when I get back home. 

Have you looked into contacting customer support? I worked on a team that racked up x100 in AWS, by mistake and got pardoned.

**Reply by AndrewLester** [reply]
Yes I think I was getting something similar in build mode locally (it didn't go infinitely)... I wasn't sure why. I assure you, if you deploy to Vercel it _will_ go infinitely 😅 (I don't recommend).

Yeah I should go about that, I think they've been bogged down with issues recently so I'm not sure how responsive they'll be. It's good to hear that companies do pardon this stuff though, I was worried!

**Comment by AndrewLester** [comment]
I'm not sure if "bumping" posts like this is against the code of conduct (I can remove), but we've received no info from Vercel support about this and are still struggling to understand the cause. Would be nice to get another set of eyes on it to help figure out the issue.

**Comment by shahsuvarli** [comment]
I have the same issue, This happened to me when I updated the api url from `/api/products/` to `http://localhost:3000/api/products`
